### ISSUE TITLE
Is dead-code elimination supported?

### ISSUE THREAD

**Issue by rodlogic** [cincheo/jsweet#223]
I am wondering if dead-code elimination is supported. The usecase I have in mind is depending on a number of candies that would go through a dead-code elimination phase before generating the final JS code for the app.

**Comment by renaudpawlak** [comment]
JSweet does not do any dead-code elimination. (but of course Javac will detect dead code and report the appropriate problem)
Could you be more specific (maybe with code samples) on how this feature would be useful?

**Comment by lgrignon** [comment]
Dead-Code elimination is often a good thing in production indeed, there are some tools we could use to apply tree shaking to bundle generation. Webpack and RollupJS both provides a builtin opportunity to do this, but browserify, that we currently use doesn't. Nevertheless, I found the following extension:
https://github.com/nolanlawson/rollupify
Would it suits your needs?

I think you could apply more complex optimizations on the generated JS with Google Closure but I never tried.

**Comment by renaudpawlak** [comment]
Note that if your use case is similar to http://stackoverflow.com/questions/4756462/java-dead-code-elimination, it would be quite easy to extend the transpiler to eliminate such dead code at compile type. JSweet 2 (factory API) will support that sort of extensions.

**Comment by rodlogic** [comment]
(Note that I have little experience with JSweet)

The Candies modularization through maven artifacts in JSweet is a killer feature IMO. However, it could also increase the size of the final assets - specially if the application relies of a larger number of candies and these candies have functionality that is not used at all. So being able to remove unused code paths when building the final application would not only keep the asset size under control but also reduce it in many cases.

**Comment by renaudpawlak** [comment]
This use case you are talking about it quite different from the one I had in mind. I believe that what you are talking about is out of scope of the transpiler. More specifically, if you care of the size of deployed JavaScript on the client side you should take a look at JavaScript modules, which can load the required code on demand (and not load the one that is not required so that you can keep your application as light as possible). JSweet supports modules with the "module" option.

Basically, regarding this issue, JSweet will support what TypeScript supports (it delegates to tsc). So you can expect:

- module support
- maybe other dead code elimination support (but for a different purpose I believe) that might be soon supported by TypeScript
 
The kind of code elimination you are talking about reminds me of a feature of GWT, which IMO is a bad/wrong alternative to modules.
